### PR TITLE
feat(gateway) add kuma Gateway types to k8s

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

This PR only adds the Kuma-internal Gateway types to k8s. Not super happy with the helm chart situation, since using Helm to install the chart would lead to the CRDs being installed in any case, even though via `install control-plane` they're skipped.

### CODE REVIEW

File: app/kumactl/cmd/install/render_helm_files.go [hunk 1]
```diff
@@ -17,6 +17,28 @@ import (
...
+type fileFilter func(data.File) bool
+
+var helmTemplateFilters []fileFilter
+
+func filterHelmTemplates(files data.FileList) data.FileList {
+	var filteredFiles data.FileList
+	for _, template := range files {
+		var filtered bool
+		for _, f := range helmTemplateFilters {
+			filtered = filtered || !f(template)
```
Comment (jpeach) [inline app/kumactl/cmd/install/render_helm_files.go:29]: "filtered" the local bool is used to mean "filter out this element", but the filters in "helmTemplateFilters" are used to mean "filter in this element". So we should either use different words or align the meanings.

File: app/kumactl/cmd/install/render_helm_files.go [hunk 2]
```diff
@@ -78,10 +110,10 @@ func renderHelmFiles(
 
 func loadCharts(templates []data.File) (*chart.Chart, error) {
 	files := []*loader.BufferedFile{}
-	for _, f := range templates {
+	for _, template := range filterHelmTemplates(templates) {
```
Comment (jpeach) [inline app/kumactl/cmd/install/render_helm_files.go:113]: I think that we should take a consistent approach to filtering the CRD files. In `install_crds.go` we filter at the top level before any subsequent processing, so why don't we just do the same thing for installing the control plane?

File: app/kumactl/cmd/install/render_helm_files_gateway.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+func init() {
+	gatewayEnabled = true
+	helmTemplateFilters = []fileFilter{
+		func(f data.File) bool {
+			return true
+		},
+	}
+}
```
Comment (jpeach) [inline app/kumactl/cmd/install/render_helm_files_gateway.go:17]: You don't need this file at all. If `helmTemplateFilter` is an empty slice, all the files will be accepted.

File: deployments/charts/kuma/values.yaml [hunk 4]
```diff
@@ -193,6 +193,10 @@ ingress:
     kubernetes.io/os: linux
     kubernetes.io/arch: amd64
 
+gateway:
+  # -- Whether the control plane is using an image with built-in Gateway support enabled
+  enabled: false
+
```
Comment (jpeach) [inline deployments/charts/kuma/values.yaml:199]: It's not all that clear to me what policy you have in mind here. Are you expecting that people would need to be able to choose to not install gateway CRDs?

| | kumactl | kuma-cp | Behaviour |
| --- | --- | --- | --- |
| Gateway Enabled | Y | Y | kumactl installs CRDs and kuma-cp works with them|
| Gateway Enabled | Y | N | kumactl installs CRDs but kuma-cp pays no attention |
| Gateway Enabled | N | Y | kumactl doesn't install CRDs and kuma-cp fails to watch the GVK (hopefully not fatal) |
| Gateway Enabled | N | N | No gateway support anywhere, it's all cool |

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_route_types_helpers.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+			Kind:       "GatewayRoute",
+		},
+	})
+	registry.RegisterListType(&mesh_proto.GatewayRoute{}, &GatewayRouteList{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: GroupVersion.String(),
+			Kind:       "GatewayRouteList",
+		},
+	})
+}
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_route_types_helpers.go:63]: Suggest that we do the same as the gateway plugin - the code is always compiled (i.e. all these methods are always present), but the types are only registered when the build flag is set.

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_types.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+// +kubebuilder:object:root=true
+type GatewayList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Gateway `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&Gateway{}, &GatewayList{})
+}
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_types.go:47]: Do you think that we should only add these to the default scheme when the build flag is set?

File: app/kumactl/cmd/install/render_helm_files_no_gateway.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"strings"
+
+	"github.com/kumahq/kuma/app/kumactl/pkg/install/data"
+)
+
+func init() {
+	helmTemplateFilters = []fileFilter{
+		func(f data.File) bool {
+			return !strings.Contains(f.FullPath, "crd") || !strings.Contains(f.Name, "gateway")
```
Comment (jpeach) [inline app/kumactl/cmd/install/render_helm_files_no_gateway.go:15]: So one idea that occured to me is to unmarshal the CRD into a `CustomResourceDefinition`, then to verify whether the kind is in the global type registry. That would automatically filter gateway types out correctly with no special cases. WDYT?

File: deployments/charts/kuma/values.yaml (hunk 4 above)
Comment (jpeach) [reply deployments/charts/kuma/values.yaml:199]: Or maybe the behavior of just always installing the CRDs doesn't seem so bad?

File: app/kumactl/cmd/install/render_helm_files.go [hunk 8]
```diff
@@ -17,6 +17,28 @@ import (
 	"github.com/kumahq/kuma/app/kumactl/pkg/install/data"
 )
 
+type fileFilter func(data.File) bool
+
+var helmTemplateFilters []fileFilter
+
+func filterHelmTemplates(files data.FileList) data.FileList {
```
Comment (jpeach) [inline app/kumactl/cmd/install/render_helm_files.go:24]: What if you updated the InstallCrdTemplateFiles hook on the InstallCrdsContext, and the InstallCpTemplateFiles hook on the other one?

File: deployments/charts/kuma/values.yaml (hunk 4 above)
Comment (michaelbeaumont) [reply deployments/charts/kuma/values.yaml:199]: Ultimately I went through this effort to avoid changing the output of `install control-plane` at all if gateway isn't enabled. Definitely open to changing that

File: app/kumactl/cmd/install/render_helm_files_gateway.go (hunk 3 above)
Comment (michaelbeaumont) [reply app/kumactl/cmd/install/render_helm_files_gateway.go:17]: 🤦

File: pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_types.go (hunk 6 above)
Comment (michaelbeaumont) [reply pkg/plugins/resources/k8s/native/api/v1alpha1/gateway_types.go]: yes

File: deployments/charts/kuma/values.yaml (hunk 4 above)
Comment (michaelbeaumont) [reply deployments/charts/kuma/values.yaml:199]: Removed the Helm value here, now `gateways` appear in the RBAC resource.

File: app/kumactl/cmd/install/render_helm_files.go (hunk 8 above)
Comment (michaelbeaumont) [reply app/kumactl/cmd/install/render_helm_files.go]: This would require updating it anywhere else a different `Install*Context` is used.

File: deployments/charts/kuma/values.yaml (hunk 4 above)
Comment (michaelbeaumont) [reply deployments/charts/kuma/values.yaml]: The issue I see is that next release CRDs will appear to users which aren't actually supported

Comment (jpeach) [reply deployments/charts/kuma/values.yaml]: > The issue I see is that next release CRDs will appear to users which aren't actually supported

Yeh, I think you are right, we should avoid that. It would be surprising to users and we could accidentally cause compatibility issues for ourselves.

File: test/framework/k8s_cluster.go [hunk 9]
```diff
@@ -422,6 +422,8 @@ func (c *K8sCluster) genValues(mode string, opts *kumaDeploymentOptions) map[str
 		values["cni.confName"] = cniCfg.confName
 	}
 
+	values["gateway.enabled"] = "true"
+
```
Comment (jpeach) [inline test/framework/k8s_cluster.go:426]: Is this change still needed now that you dropped it from the helm valued??

File: pkg/plugins/resources/k8s/native/controllers/doc.go [hunk 10]
```diff
@@ -14,3 +14,5 @@ package controllers
 // +kubebuilder:rbac:groups=kuma.io,resources=retries,verbs=get;list;watch
 
 // +kubebuilder:rbac:groups=kuma.io,resources=proxytemplates,verbs=get;list;watch;create;update;patch;delete
+
+// +kubebuilder:rbac:groups=kuma.io,resources=gateways;gatewayroutes,verbs=get;list;watch;create;update;patch;delete
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/controllers/doc.go:18]: nit: suggest we do one type per line, since that's consistent with the other definitions here, and makes it easier to see there's multiple types

File: pkg/plugins/resources/k8s/native/config/rbac/role.yaml [hunk 11]
```diff
@@ -46,6 +46,19 @@ rules:
...
+  - gatewayroutes
+  - gateways
+  verbs:
+  - create
+  - delete
+  - get
+  - list
+  - patch
+  - update
+  - watch
```
Comment (jpeach) [inline pkg/plugins/resources/k8s/native/config/rbac/role.yaml:61]: ISTR that the RBAC controller is OK with unknown API resources, so this ought to be OK when gateway types are not installed. It would be cleaner if we didn't install them, but that would end up with more build complexity.

WDYT?

**Review by jpeach — APPROVED** [review]
Overall I think this looks pretty clean. I agree with you that we should not install gateway CRDs unless the build flag is set.

File: pkg/plugins/resources/k8s/native/config/rbac/role.yaml (hunk 11 above)
Comment (michaelbeaumont) [reply pkg/plugins/resources/k8s/native/config/rbac/role.yaml:61]: Yeah I think it's an OK trade off. There's still one problem though, that installing the chart via Helm leads to the new CRDs being installed, and we can't exclude them via templates.

How about adding a step to `tools/releases/helm.sh` to delete those 2 CRDs while packaging? Not ideal perhaps. WDYT @jakubdyszkiewicz

Comment (jpeach) [reply pkg/plugins/resources/k8s/native/config/rbac/role.yaml:72]: The only clean-ish way I can see to avoid installing the gateway CRDs here is to keep them in a different source path. That would mean adding some sort of parallel helper for `deployments.KumaChartFS()` - seems doable.

We could merge this and tackle it in a followup PR if there's no risk of getting caught up in a release tag by accident.

Comment (michaelbeaumont) [reply pkg/plugins/resources/k8s/native/config/rbac/role.yaml:72]: OK, merging as is.
